data/account/participationRegistry.go

```diff
@@ -0,0 +1,947 @@
+
+			voting BLOB --*  msgpack encoding of ParticipationAccount.voting
+		)`
+
+	createStateProof = `CREATE TABLE StateProofKeys (
+			pk    INTEGER NOT NULL, --* join with keyset to find key for a particular participation id
+			round INTEGER NOT NULL, --*  committed round for this key
+			key   BLOB    NOT NULL, --*  msgpack encoding of ParticipationAccount.BlockProof.SignatureAlgorithm
+			PRIMARY KEY (pk, round)
+		)`
+	insertKeysetQuery  = `INSERT INTO Keysets (participationID, account, firstValidRound, lastValidRound, keyDilution, vrf) VALUES (?, ?, ?, ?, ?, ?)`
+	insertRollingQuery = `INSERT INTO Rolling (pk, voting) VALUES (?, ?)`
+
+	// SELECT pk FROM Keysets WHERE participationID = ?
+	selectPK      = `SELECT pk FROM Keysets WHERE participationID = ? LIMIT 1`
+	selectLastPK  = `SELECT pk FROM Keysets ORDER BY pk DESC LIMIT 1`
+	selectRecords = `SELECT
+			k.participationID, k.account, k.firstValidRound,
+       		k.lastValidRound, k.keyDilution, k.vrf,
+			r.lastVoteRound, r.lastBlockProposalRound, r.lastStateProofRound,
+			r.effectiveFirstRound, r.effectiveLastRound, r.voting
+		FROM Keysets k
+		INNER JOIN Rolling r
+		ON k.pk = r.pk`
+	deleteKeysets          = `DELETE FROM Keysets WHERE pk=?`
+	deleteRolling          = `DELETE FROM Rolling WHERE pk=?`
+	updateRollingFieldsSQL = `UPDATE Rolling
+		 SET lastVoteRound=?,
+		     lastBlockProposalRound=?,
+		     lastStateProofRound=?,
+		     effectiveFirstRound=?,
+		     effectiveLastRound=?
+		 WHERE pk IN (SELECT pk FROM Keysets WHERE participationID=?)`
+)
+
+// dbSchemaUpgrade0 initialize the tables.
+func dbSchemaUpgrade0(ctx context.Context, tx *sql.Tx, newDatabase bool) error {
+	// Keysets is for the immutable data.
+	_, err := tx.Exec(createKeysets)
+	if err != nil {
+		return err
+	}
+
+	// Rolling may change over time.
+	_, err = tx.Exec(createRolling)
+	if err != nil {
+		return err
+	}
+
+	// For performance reasons, state proofs are in a separate table.
+	_, err = tx.Exec(createStateProof)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// participationDB is a private implementation of ParticipationRegistry.
+type participationDB struct {
+	cache map[ParticipationID]ParticipationRecord
+
+	// dirty marked on Record(), cleared on Register(), Delete(), Flush()
+	dirty map[ParticipationID]struct{}
+
+	log   logging.Logger
+	store db.Pair
+	mutex deadlock.RWMutex
+
+	writeQueue chan partDBWriteRecord
+	asyncErr   error
+
+	flushDone      *sync.Cond
+	flushesPending int
+}
+
+type updatingParticipationRecord struct {
+	ParticipationRecord
+
+	required bool
+}
+
+type partDBWriteRecord struct {
+	insertID ParticipationID
+	insert   Participation
+
+	registerUpdated map[ParticipationID]updatingParticipationRecord
+
+	delete ParticipationID
+
+	flush bool
+}
+
+func (db *participationDB) initializeCache() error {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+
+	db.cache = make(map[ParticipationID]ParticipationRecord)
+	db.dirty = make(map[ParticipationID]struct{})
```

Comment:
as a defensive coding, I'd like to suggest that you'll defer the assignments of db.cache/db.dirty until after the for-loop. In case of an error, there is no reason for these to be populated.

Reply:
I think it's clearer this way, I don't think we should optimize an error case for the garbage collector at the expense of readability. Maybe if we expected errors to be hit frequently, but this would cause the system to stop anyway so I don't think there's any benefit.

Reply:
I don't think about this one as an optimization. I was thinking of having the function making no changes in the case of an error. ( i.e. so that the state hasn't changed )

Reply:
Ah, got it, yes that makes more sense to me. So I'll create a temporary variable and update everything at the end. Otherwise, on an error, the cache has an inconsistent state.